dice-game/service/src/main.rs

```diff
@@ -0,0 +1,153 @@
+use map_vec::Map;
+use oasis_std::{Address, Context};
+use rand::Rng;
+
+#[derive(oasis_std::Service)]
+struct DiceGame {
+    max_score: u32,
+    num_players: u32,
+    scores: Map<Address, u32>,
+}
+
+type Result<T> = std::result::Result<T, String>;
+impl DiceGame {
+    pub fn new(_ctx: &Context, num_players: u32) -> Self {
+        Self {
+            max_score: 0,
+            num_players,
+            scores: Map::new(),
+        }
+    }
+
+    /// returns whether still in play
+    pub fn is_in_play(&self, _ctx: &Context) -> bool {
+        self.scores.len() < (self.num_players as usize)
+    }
+
+    /// returns num_players
+    pub fn num_players(&self, _ctx: &Context) -> u32 {
+        self.num_players
+    }
+
+    /// returns the max score in play
+    pub fn max_score(&self, _ctx: &Context) -> u32 {
+        self.max_score
+    }
+
+    /// rolls a dice, resulting in a random number from 1-6
+    pub fn roll(&mut self, ctx: &Context) -> Result<u32> {
+        if self.scores.len() >= (self.num_players as usize) {
+            return Err("Maximum number of players rolled".to_string());
+        }
+
+        // each player is only allowed to roll once
+        if self.scores.contains_key(&ctx.sender()) {
+            return Err("Repeat rolls not allowed!".to_string());
+        }
+
+        let score = rand::thread_rng().gen_range(1, 7);
+        if score > self.max_score {
+            self.max_score = score;
+        }
+
+        self.scores.insert(ctx.sender(), score);
+        Ok(score)
+    }
+
+    /// returns a vector of players with the highest dice roll
+    pub fn winner(&self, _ctx: &Context) -> Result<Vec<Address>> {
+        Ok(self
+            .scores
+            .clone()
```

Comment:
I think that you might just want to do [iter().max](https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.max)

Reply:
hmm In this case Im evaluating by value but I want a vector of keys in the end, which was why I used the current approach - would used `.max` work in this use case?

Reply:
https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.max_by_key

Reply:
hmm i noticed that `max_by_key` returns only 1 element if there are multiple items that tie for max but in this case the  behavior that's needed is to return all items (multiple) that tie for max